lib/python/rose/popen.py

```diff
@@ -271,6 +272,57 @@ def which(name):
             if os.access(file_name, os.F_OK | os.X_OK):
                 return file_name
 
+    async def run_bg_async(self, *args, **kwargs):
+        """Provide a Rose-friendly interface to subprocess.Popen.
+
+        Return a subprocess.Popen object.
+
+        If kwargs["stdin"] is a str, turn it into subprocess.PIPE.
+        However, it cannot communicate stdin to the Popen object,
+        because the Popen.communicate() method implies a wait().
+
+        """
+        for key in ["stdout", "stderr"]:
+            if kwargs.get(key) is None:
+                kwargs[key] = asyncio.subprocess.PIPE
+        stdin = kwargs.get("stdin")
+        if isinstance(stdin, str):
+            kwargs["stdin"] = asyncio.subprocess.PIPE
+        elif stdin is None:
+            kwargs["stdin"] = open(os.devnull)
+        self.handle_event(RosePopenEvent(args, stdin))
+        sys.stdout.flush()
+        try:
+            if kwargs.get("shell"):
+                proc = Popen(args[0], **kwargs)
+            else:
+                command = ' '.join(args)
+                proc = await asyncio.create_subprocess_shell(command, **kwargs)
+        except OSError as exc:
+            if exc.filename is None and args:
+                exc.filename = args[0]
+            raise RosePopenError(args, 1, "", str(exc))
+        return proc
+
+    async def run_async(self, *args, **kwargs):
+        """Provide a Rose-friendly interface to subprocess.Popen.
+
+        Return ret_code, out, err.
+
+        If kwargs["stdin"] is a str, communicate it to the command via a pipe.
+
+        """
+        proc = await self.run_bg_async(*args, **kwargs)
+        stdin = None
+        if isinstance(kwargs.get("stdin"), str):
+            stdin = kwargs.get("stdin")
+        stdout, stderr = await proc.communicate(stdin)
+        return stdout, stderr
```

Comment:
Should this return `(ret_code, stdout, stderr)` for consistency with `run()`?

Reply:
I'm not sure if `ret_code` actually used. If it is, then this code is not tested properly, because putting it in breaks the tests.

Reply:
`ret_code` isn't used in this new code which makes me think that a subprocess failure won't get caught?

I think `ret_code` it should be checked in `run_ok_async` to mirror the synchronous implementation.

Reply:
Yes, I'm a bit slow today. I've now fixed this. :smile: :sleeping: